File: frontend/src/types/analysis.ts
export type ThreatClassification =
  | "Benign"
  | "Spam"
  | "Suspicious"
  | "Phishing"
  | "Malware"
  | "Credential Theft"
  | "Business Email Compromise"
  | "Data Exfiltration"
  | "Unknown"

export type RiskLevel = "Low" | "Medium" | "High" | "Critical"

export interface IOCResult {
  ips: string[]
  domains: string[]
  urls: string[]
  emails: string[]
  hashes: string[]
}

export interface MitreTechniqueDetail {
  id: string
  name: string
  tactic_id: string
  tactic_name: string
  description: string
}

export interface MitreMatrixTactic {
  id: string
  name: string
  techniques: MitreTechniqueDetail[]
}

export interface MitreMatrixResponse {
  tactics: MitreMatrixTactic[]
}

export interface ExplainabilityItem {
  label: string
  matched: boolean
}

export interface RecommendationsGrouped {
  immediate: string[]
  investigate: string[]
  contain: string[]
  recover: string[]
}

export interface SigmaMatch {
  rule_name: string
  matched: boolean
  matched_indicators: string[]
}

export interface DetectionMetrics {
  detection_confidence: number
  malicious_probability: number
  suspicious_probability: number
  false_positive_probability: number
}

export type ThreatIntelVerdict = "Malicious" | "Suspicious" | "Harmless" | "Unknown"

export interface ThreatIntelFinding {
  source: "VirusTotal" | "Shodan" | "AbuseIPDB"
  indicator: string
  indicator_type: "ip" | "domain" | "url" | "hash"
  verdict: ThreatIntelVerdict
  summary: string
  detail_url?: string | null
  malicious_engines?: number | null
  total_engines?: number | null
  error?: string | null
  checked_at: string
  from_cache?: boolean
  country?: string | null
  asn?: string | null
  reverse_dns?: string | null
  malware_family?: string | null
  blacklist_count?: number | null
  first_seen?: string | null
  last_seen?: string | null
}

export interface LocalIocFinding {
  indicator: string
  indicator_type: "ip" | "domain" | "url" | "hash" | "email"
  risk_level: RiskLevel
  threat_score: number
  confidence: number
  ip_class?: string | null
  tld?: string | null
  hash_algorithm?: string | null
  notes: string[]
}

export interface ThreatIntelEnrichment {
  virustotal_configured: boolean
  shodan_configured: boolean
  abuseipdb_configured: boolean
  findings: ThreatIntelFinding[]
  local_findings: LocalIocFinding[]
}

export interface OwaspMapping {
  id: string
  name: string
  reason: string
}

export interface PlaybookAction {
  // The backend serializes the action name as `action`; `title`/`description`
  // are kept optional for backward compatibility with earlier frontend code.
  action?: string
  title?: string
  description?: string
  priority: "Critical" | "High" | "Medium" | "Low"
  category: string
}

export interface AnalysisResponse {
  id: string
  input_text: string
  input_type: string
  classification: ThreatClassification
  risk_level: RiskLevel
  confidence: number
  summary: string
  explanation: string
  recommendations: string[]
  indicators: string[]
  mitre_techniques: string[]
  iocs: IOCResult
  model_used: string
  latency_ms: number
  created_at: string

  // Analyzer-page enrichment (may be null/undefined for older records)
  risk_score?: number | null
  threat_tags?: string[] | null
  mitre_details?: MitreTechniqueDetail[] | null
  attack_timeline?: string[] | null
  explainability?: ExplainabilityItem[] | null
  recommendations_grouped?: RecommendationsGrouped | null
  sigma_match?: SigmaMatch | null
  detection_metrics?: DetectionMetrics | null
  threat_intel?: ThreatIntelEnrichment | null
  owasp_mappings?: OwaspMapping[] | null
  risk_factors?: string[] | null
  knowledge_sources?: string[] | null
  playbook_actions?: PlaybookAction[] | null
}

export interface AnalysisListItem {
  id: string
  input_type: string
  classification: ThreatClassification
  risk_level: RiskLevel
  confidence: number
  summary: string
  created_at: string
  risk_score?: number | null
}

export interface PaginatedAnalyses {
  total: number
  page: number
  page_size: number
  items: AnalysisListItem[]
}

export interface RiskDistributionItem {
  risk_level: string
  count: number
}

export interface ClassificationDistributionItem {
  classification: string
  count: number
}

export interface WeeklyStatItem {
  date: string
  total: number
  phishing: number
  malware: number
  critical: number
}

export interface MonthlyStatItem {
  month: string
  total: number
  phishing: number
  malware: number
  critical: number
}

export interface TopItem {
  label: string
  count: number
}

export interface StatsResponse {
  total_analyses: number
  phishing_detected: number
  malware_detected: number
  critical_alerts: number
  avg_risk_score: number
  avg_confidence: number
  risk_distribution: RiskDistributionItem[]
  classification_distribution: ClassificationDistributionItem[]
  weekly_stats: WeeklyStatItem[]
  analyses_today: number
  avg_latency_ms: number
  false_positive_rate: number
  monthly_stats: MonthlyStatItem[]
  top_ioc_types: TopItem[]
  top_countries: TopItem[]
  top_malware_families: TopItem[]
  top_mitre_techniques: TopItem[]
  top_owasp_categories: TopItem[]
}

export interface AnalyzeRequest {
  content: string
  input_type?: string
}

export interface CopilotChatMessage {
  role: "user" | "assistant"
  content: string
}

/** Structured snapshot of the analysis currently on screen, sent to the
 * copilot so it can ground its answer in the actual incident. Mirrors the
 * fields called out in the SOC Copilot spec. All optional so it degrades
 * gracefully for older records that lack enrichment. */
export interface CopilotIncidentContext {
  analysisId?: string | null
  verdict?: string | null
  severity?: string | null
  riskScore?: number | null
  confidence?: number | null
  threatCategories?: string[] | null
  executiveSummary?: string | null
  detailedExplanation?: string | null
  keyIndicators?: string[] | null
  explainableSignals?: string[] | null
  mitreTechniques?: string[] | null
  iocs?: IOCResult | null
  originalContent?: string | null
  sender?: string | null
  urls?: string[] | null
  domains?: string[] | null
}

export interface CopilotChatRequest {
  message: string
  history: CopilotChatMessage[]
  analysis_id?: string | null
  incident_context?: CopilotIncidentContext | null
}

export interface CopilotChatResponse {
  reply: string
  grounded_in_analysis_id?: string | null
}

export interface HistoryQueryParams {
  page?: number
  page_size?: number
  classification?: string
  risk_level?: string
  search?: string
  sort_by?: "created_at" | "confidence" | "risk_score" | "classification" | "risk_level"
  sort_order?: "asc" | "desc"
}

File: backend/app/schemas/analysis.py
"""Pydantic schemas for the analysis endpoints and the structured LLM output."""
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


class ThreatClassification(str, Enum):
    BENIGN = "Benign"
    SPAM = "Spam"
    SUSPICIOUS = "Suspicious"
    PHISHING = "Phishing"
    MALWARE = "Malware"
    CREDENTIAL_THEFT = "Credential Theft"
    BEC = "Business Email Compromise"
    DATA_EXFILTRATION = "Data Exfiltration"
    UNKNOWN = "Unknown"


class RiskLevel(str, Enum):
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"
    CRITICAL = "Critical"


class AnalyzeRequest(BaseModel):
    content: str = Field(
        ..., min_length=3, max_length=20000, description="Raw email, alert, or log text to analyze."
    )
    input_type: Optional[str] = Field(
        default="unknown",
        description="Hint: email | soc_alert | windows_log | linux_log | other",
    )

    @field_validator("content")
    @classmethod
    def strip_content(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("content must not be empty")
        return v


class IOCResult(BaseModel):
    ips: list[str] = Field(default_factory=list)
    domains: list[str] = Field(default_factory=list)
    urls: list[str] = Field(default_factory=list)
    emails: list[str] = Field(default_factory=list)
    hashes: list[str] = Field(default_factory=list)


class LLMAnalysisResult(BaseModel):
    """The strict JSON schema the LLM must return."""

    classification: ThreatClassification
    risk_level: RiskLevel
    confidence: int = Field(..., ge=0, le=100)
    summary: str
    explanation: str
    indicators: list[str] = Field(default_factory=list)
    mitre_techniques: list[str] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)


class MitreTechniqueDetail(BaseModel):
    """MITRE ATT&CK technique enriched with tactic + description from the
    local knowledge base (not trusted from the LLM directly)."""

    id: str
    name: str
    tactic_id: str
    tactic_name: str
    description: str


class ExplainabilityItem(BaseModel):
    label: str
    matched: bool


class OwaspMapping(BaseModel):
    """OWASP Top 10 (2021) category mapped from deterministic explainability
    signals/classification -- narrative context, not a claim that this is a
    web-app vulnerability scan (see app/services/enrichment_service.py)."""

    id: str
    name: str
    reason: str


class PlaybookAction(BaseModel):
    """A single recommended SOC response action with a priority, generated
    deterministically from the threat category + severity of the analysis."""

    action: str
    priority: str  # "Critical" | "High" | "Medium" | "Low"
    category: str  # "Containment" | "Eradication" | "Forensics" | "Investigation" | "Communication"


class RecommendationsGrouped(BaseModel):
    immediate: list[str] = Field(default_factory=list)
    investigate: list[str] = Field(default_factory=list)
    contain: list[str] = Field(default_factory=list)
    recover: list[str] = Field(default_factory=list)


class SigmaMatch(BaseModel):
    rule_name: str
    matched: bool
    matched_indicators: list[str] = Field(default_factory=list)


class DetectionMetrics(BaseModel):
    detection_confidence: int
    malicious_probability: int
    suspicious_probability: int
    false_positive_probability: int


class LocalIocFinding(BaseModel):
    """Zero-dependency local enrichment for a single IOC, computed without any
    external API key. Always populated (graceful fallback per requirement),
    independent of whether VirusTotal/Shodan/AbuseIPDB are configured."""

    indicator: str
    indicator_type: str  # "ip" | "domain" | "url" | "hash" | "email"
    ip_class: Optional[str] = None
    hash_algorithm: Optional[str] = None
    tld: Optional[str] = None
    risk_level: RiskLevel
    threat_score: int = Field(..., ge=0, le=100)
    confidence: int = Field(..., ge=0, le=100)
    notes: list[str] = Field(default_factory=list)


class ThreatIntelVerdict(str, Enum):
    MALICIOUS = "Malicious"
    SUSPICIOUS = "Suspicious"
    HARMLESS = "Harmless"
    UNKNOWN = "Unknown"


class ThreatIntelFinding(BaseModel):
    """A single live lookup result for one IOC against one external source."""

    source: str  # "VirusTotal" | "Shodan" | "AbuseIPDB"
    indicator: str
    indicator_type: str  # "ip" | "domain" | "url" | "hash"
    verdict: ThreatIntelVerdict = ThreatIntelVerdict.UNKNOWN
    summary: str
    detail_url: Optional[str] = None
    malicious_engines: Optional[int] = None
    total_engines: Optional[int] = None
    error: Optional[str] = None
    checked_at: datetime
    from_cache: bool = False
    # --- Advanced IOC enrichment (additive; best-effort per source) ---
    country: Optional[str] = None
    asn: Optional[str] = None
    reverse_dns: Optional[str] = None
    malware_family: Optional[str] = None
    blacklist_count: Optional[int] = None
    first_seen: Optional[str] = None
    last_seen: Optional[str] = None


class ThreatIntelEnrichment(BaseModel):
    """Live external threat-intelligence lookups (VirusTotal + Shodan +
    AbuseIPDB) for the IOCs extracted from this analysis. Best-effort:
    `*_configured` reflects whether an API key is set, independent of
    whether any IOC was found or any lookup succeeded."""

    virustotal_configured: bool = False
    shodan_configured: bool = False
    abuseipdb_configured: bool = False
    findings: list[ThreatIntelFinding] = Field(default_factory=list)
    # Zero-dependency local heuristics, always populated regardless of
    # whether any API key above is configured (graceful fallback).
    local_findings: list[LocalIocFinding] = Field(default_factory=list)


class AnalysisEnrichment(BaseModel):
    """Deterministic additions computed server-side on top of the raw LLM
    result (see app/services/enrichment_service.py)."""

    risk_score: int = Field(..., ge=0, le=100)
    threat_tags: list[str] = Field(default_factory=list)
    mitre_details: list[MitreTechniqueDetail] = Field(default_factory=list)
    attack_timeline: list[str] = Field(default_factory=list)
    explainability: list[ExplainabilityItem] = Field(default_factory=list)
    recommendations_grouped: RecommendationsGrouped
    sigma_match: SigmaMatch
    detection_metrics: DetectionMetrics
    # --- Enterprise SOC additions (additive) ---
    owasp_mappings: list[OwaspMapping] = Field(default_factory=list)
    risk_factors: list[str] = Field(default_factory=list)
    knowledge_sources: list[str] = Field(default_factory=list)
    playbook_actions: list[PlaybookAction] = Field(default_factory=list)


class AnalysisResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

    id: str
    input_text: str
    input_type: str
    classification: ThreatClassification
    risk_level: RiskLevel
    confidence: int
    summary: str
    explanation: str
    recommendations: list[str]
    indicators: list[str]
    mitre_techniques: list[str]
    iocs: IOCResult
    model_used: str
    latency_ms: float
    created_at: datetime

    # --- Analyzer page enrichment (additive, may be null for older rows) ---
    risk_score: Optional[int] = None
    threat_tags: Optional[list[str]] = None
    mitre_details: Optional[list[MitreTechniqueDetail]] = None
    attack_timeline: Optional[list[str]] = None
    explainability: Optional[list[ExplainabilityItem]] = None
    recommendations_grouped: Optional[RecommendationsGrouped] = None
    sigma_match: Optional[SigmaMatch] = None
    detection_metrics: Optional[DetectionMetrics] = None
    threat_intel: Optional[ThreatIntelEnrichment] = None
    owasp_mappings: Optional[list[OwaspMapping]] = None
    risk_factors: Optional[list[str]] = None
    knowledge_sources: Optional[list[str]] = None
    playbook_actions: Optional[list[PlaybookAction]] = None


class AnalysisListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    input_type: str
    classification: ThreatClassification
    risk_level: RiskLevel
    confidence: int
    summary: str
    created_at: datetime
    risk_score: Optional[int] = None


class PaginatedAnalyses(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[AnalysisListItem]
